### PR TITLE
Decoy flip flop resistance

### PR DESCRIPTION
**Pull request by djrtwo**
Integrate FMD GHOST into the fork choice rule to prevent "decoy flip flop attacks" discussed [here](https://ethresear.ch/t/decoy-flip-flop-attack-on-lmd-ghost/6001).

* Only consider an `attestation` in `on_attestation` if it is from the current or previous epoch (wrt wall clock time).
* add a few tests

Note: this currently has `bounce-attack` as base branch for easy review. Will swap to `v09x` once `bounce-attack` is merged

### CODE REVIEW

File: specs/core/0_fork-choice.md [hunk 1]
```diff
@@ -224,6 +224,9 @@ def on_block(store: Store, block: BeaconBlock) -> None:
 def on_attestation(store: Store, attestation: Attestation) -> None:
     target = attestation.data.target
 
+    # Attestations must be from the current or previous epoch
+    current_epoch = compute_epoch_at_slot(get_current_slot(store))
+    assert target.epoch in [current_epoch, current_epoch - 1 if current_epoch > GENESIS_EPOCH else GENESIS_EPOCH]
```
Comment (protolambda) [inline specs/core/0_fork-choice.md]: Just noticed that the comparison is different than `get_previous_epoch` but still valid. And we can't reuse that state helper func sadly.
I'm also wondering if the spec is saying something about a target epoch set equal to the `GENESIS_EPOCH`. Checked beacon and validator spec, but everything looks to implicitly support it. Maybe we should add a note about genesis epoch for completeness?

Comment (djrtwo) [reply specs/core/0_fork-choice.md]: done

Comment (djrtwo) [reply specs/core/0_fork-choice.md]: We skip over having to deal with the corner case for finality of source and target being equal during the first epoch in `process_justification_and_finalization` with the following.

```python
    if get_current_epoch(state) <= GENESIS_EPOCH + 1:
        return
```

As for the fork choice, we want epoch 0 attestations to be incorporated

Comment (djrtwo) [reply specs/core/0_fork-choice.md]: ah, you meant a note about genesis epoch for validator attestations/fork choice. reasonable

Comment (protolambda) [reply specs/core/0_fork-choice.md]: Yes, I think `GENESIS_EPOCH` deserves a mention in the validator and/or fork-choice spec on how to attest it. Even if not that different, reading that something isn't actually missing is still useful.

Comment (djrtwo) [reply specs/core/0_fork-choice.md]: done

**Pull request merged**